clients/src/main/java/org/apache/kafka/clients/ClientTelemetryReporter.java

```diff
@@ -0,0 +1,914 @@
+         */
+        Objects.requireNonNull(this.rawOriginalConfig, "configure() was not called before contextChange()");
+        if (!metricsContext.contextLabels().containsKey(MetricsContext.NAMESPACE)) {
+            log.warn("_namespace not found in metrics context. Metrics collection is disabled");
+            return;
+        }
+
+        collectors.forEach(MetricsCollector::stop);
+
+        if (!telemetryProvider.validate(metricsContext, rawOriginalConfig)) {
+            log.warn("Validation failed for {} context {}, skip starting collectors", telemetryProvider.getClass(), metricsContext.contextLabels());
+            return;
+        }
+
+        if (kafkaMetricsCollector == null) {
+            // Initialize the provider only once. contextChange(..) can be called more than once,
+            // but once it's been initialized and all necessary labels are present then we don't
+            // re-initialize.
+            telemetryProvider.configure(rawOriginalConfig);
+        }
+
+        telemetryProvider.contextChange(metricsContext);
+
+        if (kafkaMetricsCollector == null) {
+            initCollectors();
+        }
+    }
+
+    @Override
+    public void init(List<KafkaMetric> metrics) {
+        // metrics collector may not have been initialized (e.g. invalid context labels)
+        // in which case metrics collection is disabled
+        if (kafkaMetricsCollector != null) {
+            kafkaMetricsCollector.init(metrics);
+        }
+    }
+
+    /**
+     * Method is invoked whenever a metric is added/registered
+     */
+    @Override
+    public void metricChange(KafkaMetric metric) {
+        // metrics collector may not have been initialized (e.g. invalid context labels)
+        // in which case metrics collection is disabled
+        if (kafkaMetricsCollector != null) {
+            kafkaMetricsCollector.metricChange(metric);
+        }
+    }
+
+    /**
+     * Method is invoked whenever a metric is removed
+     */
+    @Override
+    public void metricRemoval(KafkaMetric metric) {
+        // metrics collector may not have been initialized (e.g. invalid context labels)
+        // in which case metrics collection is disabled
+        if (kafkaMetricsCollector != null) {
+            kafkaMetricsCollector.metricRemoval(metric);
+        }
+    }
+
+    @Override
+    public void close() {
+        log.debug("Stopping ClientTelemetryReporter");
+        try {
+            clientTelemetrySender.close();
+        } catch (Exception exception) {
+            log.error("Failed to close client telemetry reporter", exception);
+        }
+    }
+
+    public ClientTelemetrySender telemetrySender() {
+        return clientTelemetrySender;
+    }
+
+    private void initCollectors() {
+        kafkaMetricsCollector = new KafkaMetricsCollector(
+            TelemetryMetricNamingConvention.getClientTelemetryMetricNamingStrategy(
+                telemetryProvider.domain()));
+        collectors.add(kafkaMetricsCollector);
```

Comment:
As of now the ClientTelemetryReporter has single collector but I have kept the list so this can be extended i.e. if required to collect HostProcess metrics, metrics other than JMX, etc. Please let me know if you think that we shouldn't keep the list rather work just with KafkaMetricsCollector for now.

Reply:
If we don't plan to support mort than one collector via KIP-714, I might be best to remove the list and keep the code simple. We can change it when support for a second collector is added. Otherwise the code is unnecessary complex (and we might never add support for a second collector).

Reply:
Done.